applications/DigitalTwinApplication/custom_sensors/displacement_sensor.cpp

```diff
@@ -0,0 +1,254 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         DigitalTwinApplication/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+// System includes
+#include <cmath>
+
+// External includes
+
+// Project includes
+
+// Application includes
+#include "digital_twin_application_variables.h"
+
+// Include base h
+#include "displacement_sensor.h"
+
+namespace Kratos {
+
+/// Constructor.
+DisplacementSensor::DisplacementSensor(
+    const std::string& rName,
+    const Point& rLocation,
+    const array_1d<double, 3>& rDirection,
+    const Element& rElement,
+    const double Weight)
+    : BaseType(rName, rLocation, Weight),
+      mElementId(rElement.Id()),
+      mDirection(rDirection)
+{
+    const auto direction_norm = norm_2(mDirection);
+
+    if (std::abs(norm_2(mDirection) - 1) > 1e-6) {
```

Comment:
what does -1 and 1e-6 can be too large, as your gradients  are in range of 1e-12 and etc.

Reply:
These are not for the gradients. These are to get the direction of the displacement sensor. I don't expect the direction norms to be close to 1e-12.